batch/test/test_dag.py

```diff
@@ -0,0 +1,204 @@
+import os
+import pytest
+import re
+import requests
+
+from batch.client import BatchClient
+
+
+@pytest.fixture
+def client():
+    return BatchClient(url=os.environ.get('BATCH_URL'))
+
+
+def test_simple(client):
+    batch = client.create_batch()
+    head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+    tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[head.id])
+    status = batch.wait()
+    assert status['jobs']['Complete'] == 2
+    head_status = head.status()
+    assert head_status['state'] == 'Complete'
+    assert head_status['exit_code'] == 0
+    tail_status = tail.status()
+    assert tail_status['state'] == 'Complete'
+    assert tail_status['exit_code'] == 0
+
+
+def test_missing_parent_is_400(client):
+    try:
+        batch = client.create_batch()
+        batch.create_job('alpine:3.8', command=['echo', 'head'], parents=[100000])
+    except requests.exceptions.HTTPError as err:
+        assert err.response.status_code == 400
+        assert re.search('.*invalid parent: no job with id.*', err.response.text)
+        return
+    assert False
+
+
+def test_already_deleted_parent_is_400(client):
+    try:
+        batch = client.create_batch()
+        head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+        head_id = head.id
+        head.delete()
+        tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[head_id])
+    except requests.exceptions.HTTPError as err:
+        assert err.response.status_code == 400
+        assert re.search('.*invalid parent: no job with id.*', err.response.text)
+        return
+    assert False
+
+
+def test_dag(client):
+    batch = client.create_batch()
+    head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+    left = batch.create_job('alpine:3.8', command=['echo', 'left'], parents=[head.id])
+    right = batch.create_job('alpine:3.8', command=['echo', 'right'], parents=[head.id])
+    tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[left.id, right.id])
+    status = batch.wait()
+    assert status['jobs']['Complete'] == 4
+    for node in [head, left, right, tail]:
+        status = node.status()
+        assert status['state'] == 'Complete'
+        assert status['exit_code'] == 0
+
+
+def test_cancel_tail(client):
+    batch = client.create_batch()
+    head = batch.create_job('alpine:3.8', command=['echo', 'head'])
+    left = batch.create_job('alpine:3.8', command=['echo', 'left'], parents=[head.id])
+    right = batch.create_job('alpine:3.8', command=['echo', 'right'], parents=[head.id])
+    tail = batch.create_job('alpine:3.8', command=['echo', 'tail'], parents=[left.id, right.id])
+    tail.cancel()
+    status = batch.wait()
+    assert status['jobs']['Complete'] == 3
```

Comment:
Still seems like a race condition here.  The jobs might have all finished and the cancel did nothing.  I think a lot of tests like this will cause long-term pain for us if we're not careful.

I'm not completely sure what to do.  One option is to say complete >= 3 and tail status is complete/exit code 0 or cancelled.

I didn't re-check the others.

Reply:
Good point. I also did not consider what should happen if you cancel left but tail has already started executing. The easiest thing to do is nothing, so that's what I'm going to do. There's currently no way to cancel a whole batch or all the children of a given job, if the parent is already complete.